src/hotspot/share/opto/vtransform.hpp

```diff
@@ -0,0 +1,516 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#ifndef SHARE_OPTO_VTRANSFORM_HPP
+#define SHARE_OPTO_VTRANSFORM_HPP
+
+#include "opto/node.hpp"
+#include "opto/vectorization.hpp"
+
+// VTransform:
+// - Models the transformation of the scalar loop to vectorized loop:
+//   It is a "C2 subgraph" -> "C2 subgraph" mapping.
+// - The VTransform contains a graph (VTransformGraph), which consists
+//   many vtnodes (VTransformNode).
+// - Each vtnode models a part of the transformation, and is supposed
+//   to represent the output C2 nodes after the vectorization as closely
+//   as possible.
+//
+// This is the life-cycle of a VTransform:
+// - Construction:
+//   - From SuperWord, with the SuperWordVTransformBuilder.
+//
+// - Future Plans: optimize, if-conversion, etc.
+//
+// - Schedule:
+//   - Compute linearization of the graph, into an order that respects
+//     all edges in the graph (bailout if cycle detected).
+//
+// - Apply:
+//   - Changes to the C2 IR are only made once the "apply" method is called.
+//   - Each vtnode generates its corresponding scalar and vector C2 nodes,
+//     possibly replacing old scalar C2 nodes.
+//
+// Future Plans with VTransform:
+// - Cost model: estimate if vectorization is profitable.
+// - Optimizations: moving unordered reductions out of the loop, whih decreases cost.
+// - Pack/Unpack/Shuffle: introduce additional nodes not present in the scalar loop.
+//                        This is difficult to do with the SuperWord packset approach.
+// - If-conversion: convert predicated nodes into CFG.
```

Comment:
Note: please read this description, it explains the basic idea of the `VTransform`.